third_party/libusb/webport/src/chrome_usb/chrome-login-state-hook.js

```diff
@@ -17,149 +17,100 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
-goog.provide('GoogleSmartCard.Libusb.ChromeLoginStateHook');
+goog.provide('GoogleSmartCard.LibusbLoginStateHook');
 
-goog.require('GoogleSmartCard.Libusb.ChromeUsbBackend');
-goog.require('GoogleSmartCard.DebugDump');
+goog.require('GoogleSmartCard.LibusbProxyHook');
 goog.require('GoogleSmartCard.Logging');
-goog.require('GoogleSmartCard.RemoteCallMessage');
-goog.require('GoogleSmartCard.RequestReceiver');
-goog.require('goog.Promise');
 goog.require('goog.asserts');
-goog.require('goog.array');
-goog.require('goog.Disposable');
-goog.require('goog.iter');
 goog.require('goog.log');
-goog.require('goog.log.Logger');
-goog.require('goog.messaging.AbstractChannel');
-goog.require('goog.object');
-goog.require('goog.promise.Resolver');
 
 goog.scope(function() {
 
 const GSC = GoogleSmartCard;
 
-/**
- * This class implements a hook that hides USB devices when the app is running
- * for a user who has locked their screen. This causes PCSC Lite to release all
- * owned USB devices so that they can be accessed from the sign-in profile app.
- * @extends goog.Disposable
- * @constructor
- */
-GSC.Libusb.ChromeLoginStateHook = function() {
-  /**
-   * @type {boolean}
-   * @private
-   */
-  this.simulateDevicesAbsent_ = false;
-  /** @type {!Function} @private @const */
-  this.boundOnGotSessionState_ = this.onGotSessionState_.bind(this);
-  /** @type {!goog.promise.Resolver} @private @const */
-  this.hookIsReadyResolver_ = goog.Promise.withResolver();
-  if (chrome.loginState) {
-    chrome.loginState.getProfileType(this.onGotProfileType_.bind(this));
-  } else {
-    goog.log.warning(
-        this.logger,
-        'chrome.loginState API is not available. This app might require a ' +
-            'newer version of Chrome.');
-    this.hookIsReadyResolver_.reject();
-  }
-};
-
-const ChromeLoginStateHook = GSC.Libusb.ChromeLoginStateHook;
-
-goog.inherits(ChromeLoginStateHook, goog.Disposable);
-
-/**
- * @type {!goog.log.Logger}
- * @const
- */
-ChromeLoginStateHook.prototype.logger =
-    GSC.Logging.getScopedLogger('Libusb.LoginStateHook');
+const logger = GSC.Logging.getScopedLogger('LibusbLoginStateHook');
 
-/** @override */
-ChromeLoginStateHook.prototype.disposeInternal = function() {
-  if (chrome.loginState) {
-    chrome.loginState.onSessionStateChanged.removeListener(
-        this.boundOnGotSessionState_);
+// TODO(#429): Add event unsubscription.
```

Comment:
I haven't ported the "dispose" method yet. Let me see how simple it is to add into the current PR...

Reply:
The `Promise` constructor executes the callback synchronously, so we do have the `resolveInitializationPromise_` field initialized already at this point. I know this looks weird, and that's what `goog.Promise.withResolver()` was improving - but they recommend to switch away from `goog.Promise` these days...